### ISSUE TITLE
casadi_getint_t defined as returning int, but the code-generated function returns casadi_int

### ISSUE THREAD

**Issue by mkatliar** [casadi/casadi#2218]
`casadi_getint_t` is defined as
https://github.com/casadi/casadi/blob/c6be863a2a23b4b7d010be1419abffbe30a007e8/casadi/mem.h#L45

However, in the generated code I see definitions like following:
```c
CASADI_SYMBOL_EXPORT casadi_int pendulum_ode_n_in(void) { return 2;}

/*...*/

CASADI_SYMBOL_EXPORT casadi_functions* pendulum_ode_functions(void) {
  static casadi_functions fun = {
    pendulum_ode_incref,
    pendulum_ode_decref,
    pendulum_ode_n_in,
    pendulum_ode_n_out,
    pendulum_ode_name_in,
    pendulum_ode_name_out,
    pendulum_ode_sparsity_in,
    pendulum_ode_sparsity_out,
    pendulum_ode_work,
    pendulum_ode
  };
  return &fun;
}
```

which leads to the following compiler messages:
```
/home/kotlyar/projects/tmpc/build/test/generated_pendulum.c:241:5: warning: incompatible pointer types initializing 'casadi_getint_t' (aka 'int (*)(void)') with an expression of type 'long long (void)' [-Wincompatible-pointer-types]
    pendulum_ode_n_in,
    ^~~~~~~~~~~~~~~~~
```
and leads to undefined behavior when the function is actually called.

**Comment by jgillis** [comment]
Thanks once again for your valuable bug reports. `mem.h` is not in the unittests, so I completely missed this during `casadi_int` refactoring..

This calls for a 3.4.1...

**Comment by mkotlyar** [comment]
@jgillis I need a quick fix for this. Do you think it is a good idea to add the following lines in `mem.h`:

```c
#ifndef casadi_int
#define casadi_int long long int
#endif
```

?

**Comment by mkotlyar** [comment]
Having `casadi_int` as a macro in one place (the generated code) and as a typedef (`casadi_types.h`) can create naming conflicts.
It is probably better to change `casadi_int` to `CASADI_INT_TYPE` in the generated code.

**Comment by jgillis** [comment]
That would probably work.
Or even more simple, compile your generated code with CASADI_INT_TYPE=int

**Comment by jaeandersson** [comment]
`mem.h` cannot include `casadi_types.h` because it defeats the idea if `mem.h` being a self-contained simplified API.  The whole thing should be refactored though, and probably moved into core/runtime.

**Comment by mkotlyar** [comment]
There is still inconsistency in `mem.h` which prevents the generated code from compiling.

For example, `casadi_decompress()` is declared as taking arguments of type `int*`: https://github.com/casadi/casadi/blob/f6023226b692317576a40d8986750b2cd8779433/casadi/mem.h#L70

but is called with arguments of type `casadi_int*`: https://github.com/casadi/casadi/blob/f6023226b692317576a40d8986750b2cd8779433/casadi/mem.h#L185

**Comment by jgillis** [comment]
Both should be fixed in 3.6. O, as mem is unittested with `-Wall -Werror -Wextra` without issue. Open a new issue when still seeing problems.
